odl/trafos/fourier.py

```diff
@@ -0,0 +1,1024 @@
+
+        super().__init__(dom, ran, linear=True)
+        self._fftw_plan = None
+
+    def _conj_range(self, dom):
+        """Returned the conjugate range determined from ``dom``."""
+        # Calculate range
+        recip_grid = reciprocal(
+            dom.grid, shift=self.shifts, halfcomplex=self.halfcomplex,
+            axes=self.axes)
+
+        # Always complex space
+        ran_fspace = FunctionSpace(recip_grid.convex_hull(), ComplexNumbers())
+
+        if is_real_dtype(dom.dtype):
+            ran_dtype = _TYPE_MAP_R2C[dom.dtype]
+        else:
+            ran_dtype = dom.dtype
+
+        conj_exp = dom.conj_exponent
+        ran_dspace_type = dspace_type(ran_fspace, impl='numpy',
+                                      dtype=ran_dtype)
+        ran_dspace = ran_dspace_type(recip_grid.size, dtype=ran_dtype,
+                                     exponent=conj_exp)
+
+        ran = DiscreteLp(ran_fspace, recip_grid, ran_dspace,
+                         exponent=conj_exp)
+
+        return ran
+
+    def _call(self, x, out, **kwargs):
+        """Implement ``self(x, out)`` for pyfftw backend.
+
+        Parameters
+        ----------
+        x : domain element
+            Discretized function to be transformed
+        out : range element
+            Element to which the output is written
+
+        Notes
+        -----
+        See the `pyfftw_call` function for further options.
+
+        See also
+        --------
+        pyfftw_call : Call pyfftw backend directly
+        """
+        self._call_pyfftw(x, out, **kwargs)
+
+    def _call_numpy(self, x, out):
+        """Implement ``self(x, out)`` for numpy backend."""
+        raise NotImplementedError
+
+    def _call_pyfftw(self, x, out, **kwargs):
+        """Implement ``self(x, out)`` for pyfftw backend.
+
+        Parameters
+        ----------
+        x : domain element
+            Discretized function to be transformed
+        out : range element
+            Element to which the output is written
+        planning_effort : {'estimate', 'measure', 'patient', 'exhaustive'}
+            Flag for the amount of effort put into finding an optimal
+            FFTW plan. See the `FFTW doc on planner flags
+            <http://www.fftw.org/fftw3_doc/Planner-Flags.html>`_.
+        planning_timelimit : `float`, optional
+            Limit planning time to roughly this amount of seconds.
+            Default: `None` (no limit)
+        threads : `int`, optional
+            Number of threads to use. Default: 1
+        """
+        # TODO: Implement version using Numpy FFT
+        # TODO: Implement zero padding
+
+        # We pop some kwargs options here so that we always use the ones
+        # given during init or implicitly assumed.
+        kwargs.pop('axes', None)
+        kwargs.pop('halfcomplex', None)
+        kwargs.pop('normalise_idft', None)
+
+        # We're always modifying the input, so a copy is unavoidable
+        x_cpy = x.copy()
```

Comment:
Here we need it, yes. The input is first preprocessed in place, so we need the copy. An then it's destroyed during planning if we don't use 'estimate'. In the latter case, we don't copy, otherwise yes. I don't quite know how to circumvent this.